File: SRC/zunbdb4.f
      SUBROUTINE ZUNBDB4( M, P, Q, X11, LDX11, X21, LDX21, THETA, PHI, TAUP1, TAUP2, TAUQ1, PHANTOM, WORK, LWORK, INFO )
*
*  -- LAPACK computational routine --
*  -- LAPACK is a software package provided by Univ. of Tennessee,    --
*  -- Univ. of California Berkeley, Univ. of Colorado Denver and NAG Ltd..--
*
*     .. Scalar Arguments ..
      int                INFO, LWORK, M, P, Q, LDX11, LDX21;
*     ..
*     .. Array Arguments ..
      double             PHI(*), THETA(*);
      COMPLEX*16         PHANTOM(*), TAUP1(*), TAUP2(*), TAUQ1(*), WORK(*), X11(LDX11,*), X21(LDX21,*)
*     ..
*
*  ====================================================================
*
*     .. Parameters ..
      COMPLEX*16         NEGONE, ONE, ZERO
      PARAMETER          ( NEGONE = (-1.0D0,0.0D0), ONE = (1.0D0,0.0D0), ZERO = (0.0D0,0.0D0) )
*     ..
*     .. Local Scalars ..
      double             C, S;
      int                CHILDINFO, I, ILARF, IORBDB5, J, LLARF, LORBDB5, LWORKMIN, LWORKOPT;
      bool               LQUERY;
*     ..
*     .. External Subroutines ..
      // EXTERNAL ZLARF, ZLARFGP, ZUNBDB5, ZDROT, ZSCAL, ZLACGV, XERBLA
*     ..
*     .. External Functions ..
      double             DZNRM2;
      // EXTERNAL DZNRM2
*     ..
*     .. Intrinsic Function ..
      // INTRINSIC ATAN2, COS, MAX, SIN, SQRT
*     ..
*     .. Executable Statements ..
*
*     Test input arguments
*
      INFO = 0
      LQUERY = LWORK .EQ. -1
*
      IF( M .LT. 0 ) THEN
         INFO = -1
      ELSE IF( P .LT. M-Q .OR. M-P .LT. M-Q ) THEN
         INFO = -2
      ELSE IF( Q .LT. M-Q .OR. Q .GT. M ) THEN
         INFO = -3
      ELSE IF( LDX11 .LT. MAX( 1, P ) ) THEN
         INFO = -5
      ELSE IF( LDX21 .LT. MAX( 1, M-P ) ) THEN
         INFO = -7
      END IF
*
*     Compute workspace
*
      IF( INFO .EQ. 0 ) THEN
         ILARF = 2
         LLARF = MAX( Q-1, P-1, M-P-1 )
         IORBDB5 = 2
         LORBDB5 = Q
         LWORKOPT = ILARF + LLARF - 1
         LWORKOPT = MAX( LWORKOPT, IORBDB5 + LORBDB5 - 1 )
         LWORKMIN = LWORKOPT
         WORK(1) = LWORKOPT
         IF( LWORK .LT. LWORKMIN .AND. .NOT.LQUERY ) THEN
           INFO = -14
         END IF
      END IF
      IF( INFO .NE. 0 ) THEN
         CALL XERBLA( 'ZUNBDB4', -INFO )
         RETURN
      ELSE IF( LQUERY ) THEN
         RETURN
      END IF
*
*     Reduce columns 1, ..., M-Q of X11 and X21
*
      DO I = 1, M-Q
*
         IF( I .EQ. 1 ) THEN
            DO J = 1, M
               PHANTOM(J) = ZERO
            END DO
            CALL ZUNBDB5( P, M-P, Q, PHANTOM(1), 1, PHANTOM(P+1), 1, X11, LDX11, X21, LDX21, WORK(IORBDB5), LORBDB5, CHILDINFO )
            CALL ZSCAL( P, NEGONE, PHANTOM(1), 1 )
            CALL ZLARFGP( P, PHANTOM(1), PHANTOM(2), 1, TAUP1(1) )
            CALL ZLARFGP( M-P, PHANTOM(P+1), PHANTOM(P+2), 1, TAUP2(1) )
            THETA(I) = ATAN2( DBLE( PHANTOM(1) ), DBLE( PHANTOM(P+1) ) )
            C = COS( THETA(I) )
            S = SIN( THETA(I) )
            PHANTOM(1) = ONE
            PHANTOM(P+1) = ONE
            CALL ZLARF( 'L', P, Q, PHANTOM(1), 1, DCONJG(TAUP1(1)), X11, LDX11, WORK(ILARF) )             CALL ZLARF( 'L', M-P, Q, PHANTOM(P+1), 1, DCONJG(TAUP2(1)), X21, LDX21, WORK(ILARF) )
         ELSE
            CALL ZUNBDB5( P-I+1, M-P-I+1, Q-I+1, X11(I,I-1), 1, X21(I,I-1), 1, X11(I,I), LDX11, X21(I,I), LDX21, WORK(IORBDB5), LORBDB5, CHILDINFO )
            CALL ZSCAL( P-I+1, NEGONE, X11(I,I-1), 1 )
            CALL ZLARFGP( P-I+1, X11(I,I-1), X11(I+1,I-1), 1, TAUP1(I) )
            CALL ZLARFGP( M-P-I+1, X21(I,I-1), X21(I+1,I-1), 1, TAUP2(I) )
            THETA(I) = ATAN2( DBLE( X11(I,I-1) ), DBLE( X21(I,I-1) ) )
            C = COS( THETA(I) )
            S = SIN( THETA(I) )
            X11(I,I-1) = ONE
            X21(I,I-1) = ONE
            CALL ZLARF( 'L', P-I+1, Q-I+1, X11(I,I-1), 1, DCONJG(TAUP1(I)), X11(I,I), LDX11, WORK(ILARF) )             CALL ZLARF( 'L', M-P-I+1, Q-I+1, X21(I,I-1), 1, DCONJG(TAUP2(I)), X21(I,I), LDX21, WORK(ILARF) )
         END IF
*
         CALL ZDROT( Q-I+1, X11(I,I), LDX11, X21(I,I), LDX21, S, -C )
         CALL ZLACGV( Q-I+1, X21(I,I), LDX21 )
         CALL ZLARFGP( Q-I+1, X21(I,I), X21(I,I+1), LDX21, TAUQ1(I) )
         C = DBLE( X21(I,I) )
         X21(I,I) = ONE
         CALL ZLARF( 'R', P-I, Q-I+1, X21(I,I), LDX21, TAUQ1(I), X11(I+1,I), LDX11, WORK(ILARF) )          CALL ZLARF( 'R', M-P-I, Q-I+1, X21(I,I), LDX21, TAUQ1(I), X21(I+1,I), LDX21, WORK(ILARF) )
         CALL ZLACGV( Q-I+1, X21(I,I), LDX21 )
         IF( I .LT. M-Q ) THEN
            S = SQRT( DZNRM2( P-I, X11(I+1,I), 1 )**2 + DZNRM2( M-P-I, X21(I+1,I), 1 )**2 )
            PHI(I) = ATAN2( S, C )
         END IF
*
      END DO
*
*     Reduce the bottom-right portion of X11 to [ I 0 ]
*
      DO I = M - Q + 1, P
         CALL ZLACGV( Q-I+1, X11(I,I), LDX11 )
         CALL ZLARFGP( Q-I+1, X11(I,I), X11(I,I+1), LDX11, TAUQ1(I) )
         X11(I,I) = ONE
         CALL ZLARF( 'R', P-I, Q-I+1, X11(I,I), LDX11, TAUQ1(I), X11(I+1,I), LDX11, WORK(ILARF) )          CALL ZLARF( 'R', Q-P, Q-I+1, X11(I,I), LDX11, TAUQ1(I), X21(M-Q+1,I), LDX21, WORK(ILARF) )
         CALL ZLACGV( Q-I+1, X11(I,I), LDX11 )
      END DO
*
*     Reduce the bottom-right portion of X21 to [ 0 I ]
*
      DO I = P + 1, Q
         CALL ZLACGV( Q-I+1, X21(M-Q+I-P,I), LDX21 )
         CALL ZLARFGP( Q-I+1, X21(M-Q+I-P,I), X21(M-Q+I-P,I+1), LDX21, TAUQ1(I) )
         X21(M-Q+I-P,I) = ONE
         CALL ZLARF( 'R', Q-I, Q-I+1, X21(M-Q+I-P,I), LDX21, TAUQ1(I), X21(M-Q+I-P+1,I), LDX21, WORK(ILARF) )
         CALL ZLACGV( Q-I+1, X21(M-Q+I-P,I), LDX21 )
      END DO
*
      RETURN
*
*     End of ZUNBDB4
*
      END

File: lib/src/zunbdb4.dart
// Copyright (c) 2024 Guilherme Lepsch. All rights reserved. Use of this
// source code is governed by a BSD-style license that can be found in the
// [LICENSE file](https://github.com/lepsch/dart_lapack/blob/main/LICENSE).

import 'dart:math';

import 'package:lapack/src/blas/dznrm2.dart';
import 'package:lapack/src/blas/zdrot.dart';
import 'package:lapack/src/blas/zscal.dart';
import 'package:lapack/src/box.dart';
import 'package:lapack/src/complex.dart';
import 'package:lapack/src/matrix.dart';
import 'package:lapack/src/xerbla.dart';
import 'package:lapack/src/zlacgv.dart';
import 'package:lapack/src/zlarf.dart';
import 'package:lapack/src/zlarfgp.dart';
import 'package:lapack/src/zunbdb5.dart';

void zunbdb4(
  final int M,
  final int P,
  final int Q,
  final Matrix<Complex> X11_,
  final int LDX11,
  final Matrix<Complex> X21_,
  final int LDX21,
  final Array<double> THETA_,
  final Array<double> PHI_,
  final Array<Complex> TAUP1_,
  final Array<Complex> TAUP2_,
  final Array<Complex> TAUQ1_,
  final Array<Complex> PHANTOM_,
  final Array<Complex> WORK_,
  final int LWORK,
  final Box<int> INFO,
) {
// -- LAPACK computational routine --
// -- LAPACK is a software package provided by Univ. of Tennessee,    --
// -- Univ. of California Berkeley, Univ. of Colorado Denver and NAG Ltd..--
  final X11 = X11_.having(ld: LDX11);
  final X21 = X21_.having(ld: LDX21);
  final WORK = WORK_.having();
  final TAUP1 = TAUP1_.having();
  final TAUP2 = TAUP2_.having();
  final TAUQ1 = TAUQ1_.having();
  final PHANTOM = PHANTOM_.having();
  final THETA = THETA_.having();
  final PHI = PHI_.having();
  double C = 0, S = 0;
  int I, ILARF = 0, IORBDB5 = 0, J, LLARF, LORBDB5 = 0, LWORKMIN, LWORKOPT;
  bool LQUERY;
  final CHILDINFO = Box(0);

  // Test input arguments

  INFO.value = 0;
  LQUERY = LWORK == -1;

  if (M < 0) {
    INFO.value = -1;
  } else if (P < M - Q || M - P < M - Q) {
    INFO.value = -2;
  } else if (Q < M - Q || Q > M) {
    INFO.value = -3;
  } else if (LDX11 < max(1, P)) {
    INFO.value = -5;
  } else if (LDX21 < max(1, M - P)) {
    INFO.value = -7;
  }

  // Compute workspace

  if (INFO.value == 0) {
    ILARF = 2;
    LLARF = max(Q - 1, max(P - 1, M - P - 1));
    IORBDB5 = 2;
    LORBDB5 = Q;
    LWORKOPT = ILARF + LLARF - 1;
    LWORKOPT = max(LWORKOPT, IORBDB5 + LORBDB5 - 1);
    LWORKMIN = LWORKOPT;
    WORK[1] = LWORKOPT.toComplex();
    if (LWORK < LWORKMIN && !LQUERY) {
      INFO.value = -14;
    }
  }
  if (INFO.value != 0) {
    xerbla('ZUNBDB4', -INFO.value);
    return;
  } else if (LQUERY) {
    return;
  }

  // Reduce columns 1, ..., M-Q of X11 and X21

  for (I = 1; I <= M - Q; I++) {
    if (I == 1) {
      for (J = 1; J <= M; J++) {
        PHANTOM[J] = Complex.zero;
      }
      zunbdb5(P, M - P, Q, PHANTOM(1), 1, PHANTOM(P + 1), 1, X11, LDX11, X21,
          LDX21, WORK(IORBDB5), LORBDB5, CHILDINFO);
      zscal(P, -Complex.one, PHANTOM(1), 1);
      zlarfgp(P, PHANTOM(1), PHANTOM(2), 1, TAUP1(1));
      zlarfgp(M - P, PHANTOM(P + 1), PHANTOM(P + 2), 1, TAUP2(1));
      THETA[I] = atan2(PHANTOM[1].real, PHANTOM[P + 1].real);
      C = cos(THETA[I]);
      S = sin(THETA[I]);
      PHANTOM[1] = Complex.one;
      PHANTOM[P + 1] = Complex.one;
      zlarf('L', P, Q, PHANTOM(1), 1, TAUP1[1].conjugate(), X11, LDX11,
          WORK(ILARF));
      zlarf('L', M - P, Q, PHANTOM(P + 1), 1, TAUP2[1].conjugate(), X21, LDX21,
          WORK(ILARF));
    } else {
      zunbdb5(
          P - I + 1,
          M - P - I + 1,
          Q - I + 1,
          X11(I, I - 1).asArray(),
          1,
          X21(I, I - 1).asArray(),
          1,
          X11(I, I),
          LDX11,
          X21(I, I),
          LDX21,
          WORK(IORBDB5),
          LORBDB5,
          CHILDINFO);
      zscal(P - I + 1, -Complex.one, X11(I, I - 1).asArray(), 1);
      zlarfgp(
          P - I + 1, X11(I, I - 1), X11(I + 1, I - 1).asArray(), 1, TAUP1(I));
      zlarfgp(M - P - I + 1, X21(I, I - 1), X21(I + 1, I - 1).asArray(), 1,
          TAUP2(I));
      THETA[I] = atan2(X11[I][I - 1].real, X21[I][I - 1].real);
      C = cos(THETA[I]);
      S = sin(THETA[I]);
      X11[I][I - 1] = Complex.one;
      X21[I][I - 1] = Complex.one;
      zlarf('L', P - I + 1, Q - I + 1, X11(I, I - 1).asArray(), 1,
          TAUP1[1].conjugate(), X11(I, I), LDX11, WORK(ILARF));
      zlarf('L', M - P - I + 1, Q - I + 1, X21(I, I - 1).asArray(), 1,
          TAUP2[1].conjugate(), X21(I, I), LDX21, WORK(ILARF));
    }

    zdrot(Q - I + 1, X11(I, I).asArray(), LDX11, X21(I, I).asArray(), LDX21, S,
        -C);
    zlacgv(Q - I + 1, X21(I, I).asArray(), LDX21);
    zlarfgp(Q - I + 1, X21(I, I), X21(I, I + 1).asArray(), LDX21, TAUQ1(I));
    C = X21[I][I].real;
    X21[I][I] = Complex.one;
    zlarf('R', P - I, Q - I + 1, X21(I, I).asArray(), LDX21, TAUQ1[I],
        X11(I + 1, I), LDX11, WORK(ILARF));
    zlarf('R', M - P - I, Q - I + 1, X21(I, I).asArray(), LDX21, TAUQ1[I],
        X21(I + 1, I), LDX21, WORK(ILARF));
    zlacgv(Q - I + 1, X21(I, I).asArray(), LDX21);
    if (I < M - Q) {
      S = sqrt(pow(dznrm2(P - I, X11(I + 1, I).asArray(), 1), 2) +
          pow(dznrm2(M - P - I, X21(I + 1, I).asArray(), 1), 2));
      PHI[I] = atan2(S, C);
    }
  }

  // Reduce the bottom-right portion of X11 to [ I 0 ]

  for (I = M - Q + 1; I <= P; I++) {
    zlacgv(Q - I + 1, X11(I, I).asArray(), LDX11);
    zlarfgp(Q - I + 1, X11(I, I), X11(I, I + 1).asArray(), LDX11, TAUQ1(I));
    X11[I][I] = Complex.one;
    zlarf('R', P - I, Q - I + 1, X11(I, I).asArray(), LDX11, TAUQ1[I],
        X11(I + 1, I), LDX11, WORK(ILARF));
    zlarf('R', Q - P, Q - I + 1, X11(I, I).asArray(), LDX11, TAUQ1[I],
        X21(M - Q + 1, I), LDX21, WORK(ILARF));
    zlacgv(Q - I + 1, X11(I, I).asArray(), LDX11);
  }

  // Reduce the bottom-right portion of X21 to [ 0 I ]

  for (I = P + 1; I <= Q; I++) {
    zlacgv(Q - I + 1, X21(M - Q + I - P, I).asArray(), LDX21);
    zlarfgp(Q - I + 1, X21(M - Q + I - P, I),
        X21(M - Q + I - P, I + 1).asArray(), LDX21, TAUQ1(I));
    X21[M - Q + I - P][I] = Complex.one;
    zlarf('R', Q - I, Q - I + 1, X21(M - Q + I - P, I).asArray(), LDX21,
        TAUQ1[I], X21(M - Q + I - P + 1, I), LDX21, WORK(ILARF));
    zlacgv(Q - I + 1, X21(M - Q + I - P, I).asArray(), LDX21);
  }
}
